### PR TITLE
Add option to randomise train or vehicle colours

### PR DESCRIPTION
**Pull request by mrmagic2020**
Address #22198.

https://github.com/OpenRCT2/OpenRCT2/assets/104749673/0847cbf4-3062-4329-b8e1-d257d8006a15

https://github.com/OpenRCT2/OpenRCT2/assets/104749673/376065a8-2f6a-4770-aa00-16ea5a1b96cf

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
This is such a nice idea! I really like the execution.

**Comment by SixFrags** [comment]
Great idea!

File: src/openrct2-ui/windows/Ride.cpp [hunk 1]
```diff
@@ -334,6 +335,7 @@ static Widget _colourWidgets[] = {
     MakeWidget({ 99, 190}, { 12, 12}, WindowWidgetType::ColourBtn, WindowColour::Secondary, 0xFFFFFFFF,          STR_SELECT_ADDITIONAL_COLOUR_1_TIP           ),
     MakeWidget({119, 190}, { 12, 12}, WindowWidgetType::ColourBtn, WindowColour::Secondary, 0xFFFFFFFF,          STR_SELECT_ADDITIONAL_COLOUR_2_TIP           ),
     MakeWidget({100,  74}, {239, 12}, WindowWidgetType::Checkbox,  WindowColour::Secondary, STR_RANDOM_COLOUR                                                 ),
+    MakeWidget({139, 190}, { 80, 12}, WindowWidgetType::Button,    WindowColour::Secondary, STR_RANDOMISE_VEHICLE_COLOURS, STR_RANDOMISE_VEHICLE_COLOURS_TIP  ),
```
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/Ride.cpp:338]: I think we should consider widening the button a little. It's a snug fit for English, but I think e.g. Dutch and German will require some more space. Perhaps increase to 140px width?

Comment (mrmagic2020) [reply src/openrct2-ui/windows/Ride.cpp:338]: You're right; I haven't thought about that.
How does this look?
image

Comment (AaronVanGeffen) [reply src/openrct2-ui/windows/Ride.cpp:338]: Hmm, that doesn't look great for English, does it... Perhaps we should go somewhere in between, then? 😕 Not 80px or 140px, but say, 110px?

Comment (mrmagic2020) [reply src/openrct2-ui/windows/Ride.cpp:338]: image
Looks alright to me at 110px 🧐

Comment (AaronVanGeffen) [reply src/openrct2-ui/windows/Ride.cpp:338]: Yeah, that looks like a good compromise. I actually like the extra horizontal space for English, too. Thanks.

**Review by AaronVanGeffen — APPROVED** [review]

**Comment by SpartanFrederic104** [comment]
Such a beautiful idea! Finally saves me the trouble from having to randomly select colors for all the individual cars for rides like the Bumper Cars, Flying Saucers, and Chairlift.

**Comment by AaronVanGeffen** [comment]
@Gymnasiast self-requested a review on this, but he is currently on holiday. I'm sorry to be postponing this, but also don't want to overrule him in his absence. Please wait a bit longer.

**Comment by mrmagic2020** [comment]
> @Gymnasiast self-requested a review on this, but he is currently on holiday. I'm sorry to be postponing this, but also don't want to overrule him in his absence. Please wait a bit longer.

No worries! Hope he enjoys his holiday :)

File: distribution/changelog.txt [hunk 2]
```diff
@@ -21,6 +21,7 @@
 - Feature: [#22090] [Plugin] Allow writing of paused state in non-networked settings.
 - Feature: [#22140] Add option to automatically close dropdown menus if Enlarged UI is enabled.
 - Feature: [#22150] [Plugin] Expose monthly expenditure history to the plugin API.
+- Feature: [#22206] Add option to randomise train or vehicle colours.
```
Comment (AaronVanGeffen) [inline distribution/changelog.txt:24]: Will need to be moved after review.

**Review by Gymnasiast — APPROVED** [review]
I’m always reluctant adding new "niceties", but since the implementation is clean and Aaron is enthusiastic, I will approve this.

File: src/openrct2-ui/windows/Ride.cpp [hunk 3]
```diff
@@ -4176,6 +4179,61 @@ static_assert(std::size(RatingNames) == 6);
...
+                    }
+
+                    int32_t numItems = ride->NumTrains;
+                    if (ride->vehicleColourSettings != VehicleColourSettings::perTrain)
+                        numItems = ride->num_cars_per_train;
+
+                    uint16_t i, colour;
+                    for (i = 0; i < numItems; i++)
+                    {
+                        colour = ScenarioRandMax(COLOUR_NUM_NORMAL);
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Ride.cpp:4216]: Don't use ScenarioRand(Max), this influences the game state which will cause desyncs, use UtilRand() instead.

File: src/openrct2-ui/windows/Ride.cpp [hunk 4]
```diff
@@ -4176,6 +4179,61 @@ static_assert(std::size(RatingNames) == 6);
...
+                        auto vehicleSetBodyColourAction = RideSetAppearanceAction(
+                            rideId, RideSetAppearanceType::VehicleColourBody, colour, i);
+                        GameActions::Execute(&vehicleSetBodyColourAction);
+                        if (allowChangingTrimColour)
+                        {
+                            colour = ScenarioRandMax(COLOUR_NUM_NORMAL);
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Ride.cpp:4222]: Same as above.

File: src/openrct2-ui/windows/Ride.cpp [hunk 5]
```diff
@@ -4176,6 +4179,61 @@ static_assert(std::size(RatingNames) == 6);
...
+                            auto vehicleSetTrimColourAction = RideSetAppearanceAction(
+                                rideId, RideSetAppearanceType::VehicleColourTrim, colour, i);
+                            GameActions::Execute(&vehicleSetTrimColourAction);
+                            if (allowChangingTertiaryColour)
+                            {
+                                colour = ScenarioRandMax(COLOUR_NUM_NORMAL);
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Ride.cpp:4228]: See above.

**Review by ZehMatt — CHANGES_REQUESTED** [review]
Left a few comments, otherwise the PR also seems fine to me.

**Comment by cheweytoo** [comment]
I like this a lot!

It should avoid the "special colours" though, which at the moment are used even if not enabled in the cheats:
vrt

I'd argue that it shouldn't use those colours ever, even when they are enabled.

File: src/openrct2-ui/windows/Ride.cpp [hunk 6]
```diff
@@ -4166,6 +4169,61 @@ static_assert(std::size(RatingNames) == 6);
...
+                    }
+
+                    int32_t numItems = ride->NumTrains;
+                    if (ride->vehicleColourSettings != VehicleColourSettings::perTrain)
+                        numItems = ride->num_cars_per_train;
+
+                    uint16_t i, colour;
+                    for (i = 0; i < numItems; i++)
+                    {
+                        colour = UtilRand() & COLOUR_NUM_NORMAL;
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Ride.cpp:4206]: `&` ?, `00110110` is a bit of an odd mask.

Comment (Gymnasiast) [reply src/openrct2-ui/windows/Ride.cpp:4206]:
```suggestion
                        colour = UtilRand() % COLOUR_NUM_NORMAL;
```

**Comment by Gymnasiast** [comment]
@cheweytoo https://github.com/OpenRCT2/OpenRCT2/pull/22585

File: src/openrct2-ui/windows/Ride.cpp [hunk 7]
```diff
@@ -4166,6 +4169,61 @@ static_assert(std::size(RatingNames) == 6);
...
+                            allowChangingTertiaryColour = true;
+                        }
+                    }
+
+                    int32_t numItems = ride->NumTrains;
+                    if (ride->vehicleColourSettings != VehicleColourSettings::perTrain)
+                        numItems = ride->num_cars_per_train;
+
+                    uint16_t i, colour;
+                    for (i = 0; i < numItems; i++)
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Ride.cpp:4204]: I also overlooked this, no need to keep the variable outside the loop.

**Pull request merged**